### PR TITLE
Use galcheat in BTK

### LINKED ISSUE #222: start using galcheat in BTK

### LINKED ISSUE #268: add filter wavelength from galcheat rather than reference dictionary in survey.py
Once it's ready in galcheat see temporariy fix in #228

### LINKED ISSUE #269: use flux calculations from galcheat
Once this PR is merged 

https://github.com/aboucaud/galcheat/pull/24

### PR DESCRIPTION
**Pull request by ismael-mendoza**
Closes #222 
Closes #269 
Closes #268 

- [x] run poetry update
- [x] add galcheat as a dependency in pyproject.toml
- [x] get rid of most `survey.py` functions that load surveys, etc. 
- [x] get rid of repeated YAML files
- [x] can use the hydra list config to specify list of surveys that want to be loaded from `galcheat.survey_info` dict
- [x] ~~make a new config directory + corresponding YAML files only for the PSF. PSF can be added to survey object/filters after creation~~

### CODE REVIEW

**Comment by ismael-mendoza** [comment]
I'm back to working on this issue @aboucaud - I will let you know if I have any questions or if a review might be helpful.

**Comment by ismael-mendoza** [comment]
Hi @aboucaud and @thuiop this is now ready for review. I decided to keep the central wavelenghts as a temporary dictionary for now in `btk/survey.py`. Once we figure out how they are calculated and added to galcheat we can grab them from there (it should be a pretty quick switch after that). Same with the flux calculations, I'm still using the default ones in BTK until this [PR](https://github.com/aboucaud/galcheat/pull/24) is merged.  

All survey information is now grabbed from a survey in galcheat. The psf can still be customized via a `psf_func` in `get_surveys`. Please let me know if you have any suggestions!

**Comment by thuiop** [comment]
It seems mostly good to me.
A few remarks : 
- it could be worth having some documentation on this (maybe in another PR)
- if the user wants to create his own survey, what would be the procedure ? Should he create a BTK survey directly, create a galcheat one ? This also needs to be specified somewhere, possibly in a tutorial

In any case, we should wait for Alex's review before merging.

**Comment by ismael-mendoza** [comment]
Thanks @thuiop 

> it could be worth having some documentation on this (maybe in another PR)

Yes good, this will probably deprecate some of the documentation and we need to update it. I was thinking of closing #269 and #268 before adding documentation.

> if the user wants to create his own survey, what would be the procedure ? Should he create a BTK survey directly, create a galcheat one ? This also needs to be specified somewhere, possibly in a tutorial

Actually a user can still create a Survey, Filter in the same way as before. The namedtuple classes are still inside BTK. This is also actually that is up for debate. So we keep this classes or completely supersede with galcheat ? The reason I did not do this was because galcheat classes do not have PSFs inside of them. If we decide to get rid of the internal BTK classes then we would need to (1) separate the PSF out of the filter (2) users can only customize them via galcheat (e.g. creating their own galcheat yaml)

File: btk/survey.py
Comment (aboucaud) [inline btk/survey.py:97]: This looks a bit like hacking. 
I would find it much more straightforward to use the following
```suggestion
        for band in survey.available_filters:
            filt = survey.get_filter(band)
```

File: btk/survey.py
Comment (aboucaud) [inline btk/survey.py:128]: I find the name `galcheat_psf` confusing because it kind of suggests galcheat contains a PSF.

File: btk/survey.py
Comment (aboucaud) [inline btk/survey.py:143]: Here when you will use the `effective_wavelength` from galcheat, make sure you convert it to angstroms first.

File: btk/survey.py
Comment (ismael-mendoza) [reply btk/survey.py:97]: yup that is better

File: btk/survey.py
Comment (ismael-mendoza) [reply btk/survey.py:129]: good point, perhaps `get_default_psf` was better

File: btk/survey.py
Comment (ismael-mendoza) [reply btk/survey.py:146]: I thought we wanted nanometers see documentation below? (which is the default in galcheat I think?)

Comment (ismael-mendoza) [reply btk/survey.py]: See this [line](https://github.com/LSSTDESC/BlendingToolKit/blob/cf3db28baa2d7d8ddba16ef503533dafb278b69b/btk/survey.py#L178). Actually I'm confused why this is 1e-10 and not 1e-9. Do you know @thuiop ?

**Review by aboucaud — COMMENTED** [review]
Just a comment when I see the `btk.survey.py`
Keep in mind that the Survey and Filter dataclasses from galcheat are extendable.

So say you want to attach a PSF array to it, it can be done as simple as 
```python
u_filter.psf = psf_array
```

File: btk/survey.py
Comment (thuiop) [reply btk/survey.py]: The wavelengths in _central_wavelength were in 1e-10 meters for some reason (not sure why, this comes from WLD).

Comment (ismael-mendoza) [reply btk/survey.py]: Ah I forgot the Angstrom definition, so that is why there is a 1e-10. Ok I fixed it

**Review by aboucaud — COMMENTED** [review]
For the units conversion, there is IMHO a cleaner alternative than `to(u.unit).value` => `to_value('unit')`.

```python
pix_stamp_size = int(self.stamp_size / s.pixel_scale.to(u.arcsec).value)
# would then look like 
pix_stamp_size = int(self.stamp_size / s.pixel_scale.to_value('arcsec'))
```

1. no need to import `astropy`
2. looks better :)

**Comment by ismael-mendoza** [comment]
Thanks for the suggestion @aboucaud - I didn't know this was a thing

**Comment by ismael-mendoza** [comment]
Hi @thuiop and @aboucaud I thought about the `FList` pickle issue more in galcheat (https://github.com/aboucaud/galcheat/issues/88). And I can't figure out a clean solution, here are some options I can think about.

1. Abandon the FList and use dictionaries 😢  (see https://github.com/aboucaud/galcheat/pull/92) 
2. Go back to using our own Survey/Filter classes in BTK.
3. Internally convert galcheat Survey/Filter classes to dictionaries before multiprocessing step. Then convert back to galcheat objects.

Any additional suggestions? Are you leaning towards any particular one?

**Comment by aboucaud** [comment]
> 1. Abandon the FList and use dictionaries 😢  (see [Fix pickle issue for multiprocessing aboucaud/galcheat#92](https://github.com/aboucaud/galcheat/pull/92))

Have you tried it to see if solves the issue ?

**Comment by ismael-mendoza** [comment]
@aboucaud - could you rebase that FList PR so I can test it in BTK? Thank you :)

File: btk/draw_blends.py
Comment (aboucaud) [inline btk/draw_blends.py:72]:
```suggestion
    total_flux = mag2counts(entry[filt.name + "_ab"], survey, filt).to_value("electron")
```

Concerning the fact the input mag seems to be an AB magnitude, is this a residual from the past or not. Because I though we had changed this and defaulted to classical magnitudes.

File: btk/draw_blends.py
Comment (aboucaud) [inline btk/draw_blends.py:620]:
```suggestion
            gal_flux = mag2counts(entry[mag_name], survey, filt).to_value("electron")
```

File: btk/draw_blends.py
Comment (aboucaud) [inline btk/draw_blends.py:622]:
```suggestion
            gal_flux = mag2counts(entry["ref_mag"], survey, filt).to_value("electron")
```

File: btk/draw_blends.py
Comment (aboucaud) [inline btk/draw_blends.py:423]:
```suggestion
        sky_level = mean_sky_level(survey, filt).to_value("electron")
```

File: btk/metrics.py
Comment (aboucaud) [inline btk/metrics.py:753]:
```suggestion
                        "pixel_scale": surveys[i].pixel_scale.to_value("arcsec"),
```

File: btk/metrics.py
Comment (aboucaud) [inline btk/metrics.py:757]: Seems a bit overkill to retrieve the filter name. I guess there should be a better way.

File: btk/metrics.py
Comment (aboucaud) [inline btk/metrics.py:787]:
```suggestion
                    "pixel_scale": surveys[0].pixel_scale.to_value("arcsec"),
```

File: btk/draw_blends.py
Comment (ismael-mendoza) [reply btk/draw_blends.py:620]: good catch! For some reason in my mind I read your post as two separate operations 😆

File: btk/metrics.py
Comment (ismael-mendoza) [reply btk/metrics.py:757]: the easiest would be to replace `meas_band_num` with the band name directly I think. What do you think @thuiop is that feasible in `metrics.py` ?

**Comment by ismael-mendoza** [comment]
@thuiop in the CI we are getting an error in `test_draw.py`

```
E       AssertionError: 
E       Arrays are not almost equal to 3 decimals
E       Did not get desired std of pixel values of blend images
E        ACTUAL: 403.79788119078427
E        DESIRED: 403.80063778695086
```

but not in my computer. Could there be another subtle randomization issue? (or is this something from galcheat?)

**Comment by ismael-mendoza** [comment]
> > 1. Abandon the FList and use dictionaries 😢  (see [Fix pickle issue for multiprocessing aboucaud/galcheat#92](https://github.com/aboucaud/galcheat/pull/92))
> 
> Have you tried it to see if solves the issue ?

Thanks @aboucaud I just tried it in this seems to resolve the issue.

File: btk/metrics.py
Comment (thuiop) [reply btk/metrics.py:757]: We do not really use the band name except to recover the corresponding filter. However, we do need the meas_band_num at some point, to access the images in that band.

**Comment by thuiop** [comment]
> @thuiop in the github CI we are getting an error in `test_draw.py` for python3.8
> 
> ```
> E       AssertionError: 
> E       Arrays are not almost equal to 3 decimals
> E       Did not get desired std of pixel values of blend images
> E        ACTUAL: 403.79788119078427
> E        DESIRED: 403.80063778695086
> ```
> 
> but not in my computer. Could there be another subtle randomization issue? (or is this something subtle from galcheat?)
> 
> Strangely this only happens with this test thinking Specifically the `test_default` in `test_draw.py` has several matching tests and corresponding assertions comparing values. And I just checked that they all pass except the very last assertion in `match_blend_images_default` which is the last matching test, very suspicious...

Definitely weird. This should be noise related I guess, but the other checks should not be passing then.  I do get the same error as the CI on my computer though, so we may as well change it...

**Comment by ismael-mendoza** [comment]
What version of python are you using? and OS ?

File: btk/metrics.py
Comment (ismael-mendoza) [reply btk/metrics.py:757]: Oh yeah good point, let's keep it as is then. Unless @aboucaud has an alternative suggestion?

**Comment by ismael-mendoza** [comment]
I think this is ready for review @aboucaud and @thuiop I would appreciate any additional suggestions if you have any (or otherwise feel free to approve). All tests are now passing

@thuiop could you help me out with changing the corresponding documentation and notebooks if you have time? (#270) Thanks! 

@aboucaud - do you intend to merge the FList PR from galcheat so that it now works on BTK or should we try to think of an alternative?

File: btk/draw_blends.py
Comment (aboucaud) [inline btk/draw_blends.py:514]:
```suggestion
                nx=pix_stamp_size, ny=pix_stamp_size, scale=survey.pixel_scale.to_value("arcsec")
```

File: btk/draw_blends.py
Comment (aboucaud) [inline btk/draw_blends.py:628]:
```suggestion
        pix_stamp_size = int(self.stamp_size / survey.pixel_scale.to_value("arcsec"))
```

File: btk/draw_blends.py
Comment (aboucaud) [inline btk/draw_blends.py:636]:
```suggestion
            nx=pix_stamp_size, ny=pix_stamp_size, scale=survey.pixel_scale.to_value("arcsec")
```

File: btk/metrics.py
Comment (aboucaud) [reply btk/metrics.py:757]: would have to go deeper in the code for proper suggestion. another time

File: btk/metrics.py
Comment (aboucaud) [inline btk/metrics.py:760]:
```suggestion
                        mean_sky_level(surveys[i], filtr).to_value("electron")
```

File: btk/metrics.py
Comment (aboucaud) [inline btk/metrics.py:794]:
```suggestion
                    mean_sky_level(surveys[0], filtr).to_value("electron")
```

File: btk/survey.py
Comment (aboucaud) [inline btk/survey.py:14]: You might want to add a bit more details to the docstring, like the fact that you return a list of modified galcheat.Survey instances and maybe rename the method to something like `get_survey_list`.

Comment (aboucaud) [reply btk/survey.py:14]: Plus I guess there should not be a default value for such method.

File: btk/survey.py
Comment (aboucaud) [inline btk/survey.py:46]: Very difficult to handle a function that can return both a list and a single value.
You might want to chose one return type only (maybe in a future PR)

File: btk/survey.py
Comment (aboucaud) [inline btk/survey.py:65]:
```suggestion
        survey.mirror_diameter.to_value("m"),
        survey.effective_area.to_value("m2"),
        filtr.psf_fwhm.to_value("arcsec"),
        filt_wavelength=filtr.effective_wavelength.to_value("angstrom"),
```

File: btk/survey.py
Comment (aboucaud) [inline btk/survey.py:152]:
```suggestion
        galsim.Image(psf_array), scale=survey.pixel_scale.to_value("arcsec")
```

File: conf/hydra/help/btk_help.yaml
Comment (aboucaud) [inline conf/hydra/help/btk_help.yaml:58]:
```suggestion
      btk surveys=LSST surveys.LSST.zeropoint_airmass=1.1 (...)
```

File: docs/source/cli.rst
Comment (aboucaud) [inline docs/source/cli.rst:14]:
```suggestion
    btk sampling=default draw_blends=catsim max_number=3 save_path=/directory/to/save/results cpus=1 verbose=False surveys=[LSST, COSMOS] surveys.LSST.zeropoint_airmass=1.1
```

File: docs/source/cli.rst
Comment (aboucaud) [inline docs/source/cli.rst:71]:
```suggestion
          btk surveys=LSST surveys.LSST.zeropoint_airmass=1.1 (...)
```

File: docs/source/tutorials.rst
Comment (aboucaud) [inline docs/source/tutorials.rst:176]: Is this not outdated ?

File: docs/source/tutorials.rst
Comment (aboucaud) [inline docs/source/tutorials.rst:285]: different from line 451 below, might want to homogenize

File: docs/source/tutorials.rst
Comment (aboucaud) [inline docs/source/tutorials.rst:451]: here

File: notebooks/00-intro.ipynb
Comment (aboucaud) [inline notebooks/00-intro.ipynb:231]: To be updated maybe to relate a bit more to `galcheat`.

File: notebooks/00-intro.ipynb
Comment (aboucaud) [inline notebooks/00-intro.ipynb:269]: oudated

File: notebooks/01b-scarlet-measure.ipynb
Comment (aboucaud) [inline notebooks/01b-scarlet-measure.ipynb:132]:
```suggestion
    "        psf = np.array([p.drawImage(galsim.Image(image.shape[1],image.shape[2]),scale=survey.pixel_scale.to_value("arcsec")).array for p in batch[\"psf\"][survey.name]])\n",
```

**Review by aboucaud — APPROVED** [review]
Apart from a few minor changes, and doc paragraphes to update, code is looking good.

**Comment by aboucaud** [comment]
> @aboucaud - do you intend to merge the FList PR from galcheat so that it now works on BTK or should we try to think of an alternative? I know removing it is not the ideal solution for galcheat...

I think I'm ready to drop the functionality. That was fun to write, did something new, but in the end this is probably not the most useful think to keep. Plus if we'd like to restrict as much as possible the ways to access stuff within `galcheat` to simplify the API, this is the right way to go.

File: btk/draw_blends.py
Comment (ismael-mendoza) [reply btk/draw_blends.py:236]: whoops forgot a couple, I thought I got them all. Thanks!

File: docs/source/tutorials.rst
Comment (ismael-mendoza) [reply docs/source/tutorials.rst:176]: yup we need to update doc/notebooks still in #270

**Comment by thuiop** [comment]
> I think this is ready for review @aboucaud and @thuiop I would appreciate any additional suggestions if you have any (or otherwise feel free to approve). All tests are now passing smiley
> 
> @thuiop could you help me out with changing the corresponding documentation and notebooks if you have time? (#270) Once this is merged? Thanks!
> 
> @aboucaud - do you intend to merge the FList PR from galcheat so that it now works on BTK or should we try to think of an alternative? I know removing it is not the ideal solution for galcheat...

Sure, I can take care of the doc after this is merged. I will also try to make a review tomorrow to make sure we do not miss anything.

**Comment by ismael-mendoza** [comment]
> Sure, I can take care of the doc after this is merged. I will also try to make a review tomorrow to make sure we do not miss anything.

Thank you 🙏 I'll merge sometime this morning probably

File: btk/survey.py
Comment (ismael-mendoza) [reply btk/survey.py:46]: See #280

File: docs/source/tutorials.rst
Comment (ismael-mendoza) [reply docs/source/tutorials.rst:285]: related to #280

File: notebooks/00-intro.ipynb
Comment (ismael-mendoza) [reply notebooks/00-intro.ipynb:269]: Thanks @aboucaud for pointing out where the docs/notebooks need to be updated. I linked to your comments in #270 since it will be done in a future PR

File: btk/draw_blends.py
Comment (ismael-mendoza) [reply btk/draw_blends.py:72]: The flux in galsim reads

```
    flux:               The flux (in photons/cm^2/s) of the profile. [default: 1]
```

is it ok to be using electrons?

File: btk/survey.py
Comment (ismael-mendoza) [reply btk/survey.py:16]: I agree thanks

File: btk/draw_blends.py
Comment (thuiop) [reply btk/draw_blends.py:72]: I think this really is the flux in electrons ; there may be some technicality here. Maybe we should open an issue ?

**Review by thuiop — APPROVED** [review]
I did not have much to comment on actually, it looks mostly good to me.

**Comment by ismael-mendoza** [comment]
OK I think this is ready to merge - there are some rough edges with the docs and noteboks described in #270 that we should look into next.

**Pull request merged**
